src/lib/battery/battery.h

```diff
@@ -32,59 +32,179 @@
 
 	/**
 	 * Update current battery status message.
 	 *
-	 * @param voltage_v: current voltage in V
-	 * @param current_a: current current in A
-	 * @param connected: Battery is connected
-	 * @param selected_source: This battery is on the brick that the selected source for selected_source
+	 * @param voltage_raw Battery voltage read from ADC, in Volts
+	 * @param current_raw Voltage of current sense resistor, in Amps
+	 * @param timestamp Time at which the ADC was read (use hrt_absolute_time())
+	 * @param selected_source This battery is on the brick that the selected source for selected_source
 	 * @param priority: The brick number -1. The term priority refers to the Vn connection on the LTC4417
-	 * @param throttle_normalized: throttle from 0 to 1
+	 * @param throttle_normalized Throttle of the vehicle, between 0 and 1
+	 * @param armed Arming state of the vehicle
+	 * @param should_publish If True, this function published a battery_status uORB message.
 	 */
-	void updateBatteryStatus(hrt_abstime timestamp, float voltage_v, float current_a,
-				 bool connected, bool selected_source, int priority,
-				 float throttle_normalized,
-				 bool armed, battery_status_s *status);
+	void updateBatteryStatus(hrt_abstime timestamp, float voltage_v, float current_a, bool connected,
+				 bool selected_source, int priority, float throttle_normalized, bool armed, bool should_publish);
+
+	/**
+	 * Publishes the uORB battery_status message with the most recently-updated data.
+	 */
+	void publish();
+
+	/**
+	 * Some old functionality expects the primary battery to be published on instance 0. To maintain backwards
+	 * compatibility, this function allows the advertisements (and therefore instances) of 2 batteries to be swapped.
+	 * However, this should not be relied upon anywhere, and should be considered for all intents deprecated.
+	 *
+	 * The proper way to uniquely identify batteries is by the `id` field in the `battery_status` message.
+	 */
+	void swapUorbAdvert(Battery &other);
+
+protected:
+	struct {
+		param_t v_empty;
+		param_t v_charged;
+		param_t n_cells;
+		param_t capacity;
+		param_t v_load_drop;
+		param_t r_internal;
+		param_t low_thr;
+		param_t crit_thr;
+		param_t emergen_thr;
+		param_t source;
+
+		param_t v_empty_old;
+		param_t v_charged_old;
+		param_t n_cells_old;
+		param_t capacity_old;
+		param_t v_load_drop_old;
+		param_t r_internal_old;
+		param_t source_old;
+	} _param_handles;
+
+	struct {
+		float v_empty;
+		float v_charged;
+		int n_cells;
+		float capacity;
+		float v_load_drop;
+		float r_internal;
+		float low_thr;
+		float crit_thr;
+		float emergen_thr;
+		int source;
+
+		float v_empty_old;
+		float v_charged_old;
+		int n_cells_old;
+		float capacity_old;
+		float v_load_drop_old;
+		float r_internal_old;
+		int source_old;
+	} _params;
+
+	battery_status_s _battery_status;
+
+	const int _index;
+
+	uORB::Subscription _param_sub{ORB_ID(parameter_update)};
```

Comment:
Why do you handle this in Battery? I prefer the existing mechanism via `ModuleParams`, then you don't need it here as well.

Reply:
Because I am using instanced parameters in the yaml file, I was under the impression that I wouldn't be able to use the nice ModuleParams API. Is there some way to properly use the ModuleParams API here? I tried using it in my old solution, but it was very messy and required a different class for each instance, which increased code size.

This current solution is not ideal, but I am not aware of a better way to use instanced parameters.

Reply:
You can still use it for the notification, just override `updateParams`. The rest is better done via C API at the moment, you are right.

Reply:
Oh OK, now I understand!